### PR TITLE
Benchmark report

### LINKED ISSUE #110: Automatic benchmark report
It would be good if we could produce a benchmark report automatically by some call to Holmake, say Holmake benchmark.txt. With such a facility it would be easy to measure the impact of any adjustments to the compiler implementation. I'm particularly interested in this now, because I'm about to implement call-intro and related closLang optimisations.

Initially the benchmark programs could just be the ones we've used before fib, qsort, queue and btree. Later we could add more. 

The benchmarks should of course measure the speed of the generated CakeML binary. The benchmarks file could also include other measures such as the length of the generated byte list. Later we can add more measures: number of executed machine instructions.

### PR DESCRIPTION
**Pull request by myreen**
Closes #110

### CODE REVIEW

File: compiler/benchmarks/README.md [hunk 1]
```diff
@@ -1,71 +1,9 @@
-Two benchmark suites for the CakeML compiler.
+Two benchmark suites for the CakeML compiler
```
[the remark this answers is not in the document]
Comment (myreen) [reply compiler/benchmarks/README.md]: The text is not a sentence. That's why I removed the full stop.

File: compiler/benchmarks/Holmakefile [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+DIRS = $(wildcard */)
+README.md: $(dprot $(CAKEMLDIR)/developers/readme_gen) readmePrefix $(README_SOURCES)
+	$(protect $(CAKEMLDIR)/developers/readme_gen) $(README_SOURCES)
+
+CAKEML_PLOTS = cakeml_plot1noerr.svg cakeml_plot2noerr.svg cakeml_plot3noerr.svg cakeml_all.dat
+MLTON_PLOTS = mlton_plot1noerr.svg mlton_plot2noerr.svg mlton_plot3noerr.svg mlton_all.dat
+EXTRA_CLEANS = $(wildcard *.dat *.html *.svg)
+
+cakeml_report.html: $(CAKEML_PLOTS)
+	echo '<!doctype html>' > cakeml_report.html; \
```
[the remark this answers is not in the document]
Comment (myreen) [reply compiler/benchmarks/Holmakefile]: Yes, there should probably be a program producing this html, but I won't have time to write it before I go on holiday (Sat).

**Comment by myreen** [comment]
I have installed the required python packages (`matplotlib` and `numpy`) on oven2 which means that this regression test might pass at least there. 

By using non-standard python packages (which on Ubuntu can be installed with `apt install python-matplotlib` and `apt install python-numpy`) we avoid depending on `gnuplot`.

One can argue that we should use CakeML for producing the diagrams, but until we have good drawing/plotting facilities in CakeML, I think we should use python's `matplotlib` (or `gnuplot`).

I note that most users won't build the benchmark report so this extra requirement is only for the regression machines. I think it's good to build the benchmark report on each regression test in order to make sure the report building infrastructure still works.

**Comment by tanyongkiam** [comment]
I've now made the python-dependent step optional: compilation continues but if python isn't available or any of its dependencies aren't installed then it's skipped.

Also, the Holmakefile now calls python3 instead of python. The same benchmarking script should work on both versions once the dependencies are installed though.

**Pull request merged**